File: Lua/BlackJack/blackjack/entity/Table.lua

local Player = require "entity.Player"
local Bank = require "entity.Bank"
local Dealer = require "entity.Dealer"
local inputHandler = require("input.ConsoleInputHandler"):getInstance()
local ActionType = require "type.ActionType"
local GameState = require "type.GameState"
local ResultType = require "type.ResultType"
local ClassicTableView = require "gfx.ClassicTableView"
local ClassicPlayer = require "behavior.ClassicPlayer"
local ClassicBank = require "behavior.ClassicBank"
local Calculator = require("entity.Calculator"):getInstance()


local Table = {}

setmetatable(Table, {
    __call = function(class,...)
        return class:new(...)
    end
})

function Table:new(numberOfDecks, inputHandler)

    local newTable = setmetatable({}, self)
    self.__index = self

    newTable.inputHandler = inputHandler
    newTable.dealer = Dealer(numberOfDecks, self)
    newTable.player = Player(0, newTable.dealer, ClassicPlayer(), inputHandler, self)
    newTable.bank = Bank(0, newTable.dealer, ClassicBank(), nil, self)
    newTable.state = GameState.NEW
    newTable.isLowDeck = false
    newTable.tableView = ClassicTableView(newTable.player, newTable.bank)
    newTable.wincount = 0
    newTable.loseCount = 0

    return newTable
end

function Table:startGame()
    print("\nWelcome to the table! Do not forget, the house always wins!")

    while(self.state ~= GameState.QUIT) do
        if(self.state == GameState.RUNNING) then self:play()
        elseif(self.state == GameState.NEW) then self:settings()
        elseif(self.state == GameState.PAUSED) then self:saveState()
        elseif(self.state == GameState.END) then self:showStatistics()
        elseif(self.state == GameState.QUIT) then
        end
    end

    print("Thank you for choosing our BlackJack app! \nBye-Bye! Have a nice day!")
end

function Table:settings()
    print("Please enter the amount of money you want to have!")

    local newAmount = self.inputHandler:readNumber()

    while(type(newAmount) ~= type(0)) do
        print("Incorrect value!")
        newAmount = self.inputHandler:readNumber()
    end

    self.player = newAmount

    self.winCount = 0
    self.loseCount = 0

    print("n - Start new game\tq - Quit")
    local valid = false
    while(valid == false) do

        local key = inputHandler.readKey()
        if( key == ActionType.NEW)then
            self.state = GameState.RUNNING
            valid = true
        elseif(key ==  ActionType.END) then
            self.state = GameState.QUIT
            valid = true
        else
            valid = false
        end
    end
end

function Table:showStatistics()                                 --TODO: implementálni

    print("\nResults:\n \tYou won "..self.winCount.." times\n\tYou lost "..self.loseCount.." times\n\tYour profit: "..self.bank:showMoney() * -1)

    print("n - Start new game\tq - Quit")
    local valid = false
    while(valid == false) do

        local key = inputHandler.readKey()
        if( key == ActionType.NEW)then
            self.state = GameState.NEW
            valid = true
        elseif(key ==  ActionType.END) then
            self.state = GameState.QUIT
            valid = true
        else
            valid = false
        end
    end

end

function Table:saveState()

    self.state = GameState.RUNNING                              -- TODO: Kivenni! soon: Fájlkezelés
end

function Table:play()

    if(self.isLowDeck) then
        self.dealer:createNewDeck()
        self.isLowDecl = false
    end

    self.player:preparation()
    self.bank:preparation()

    self.player:raise()

    self.player:receiveFirstCards()
    self.bank:receiveFirstCards()


    self.player:play()

    if( self.state ~= GameState.RUNNING ) then return
    end

    self.bank:play()

    self:calculateResult()

end

function Table:calculateResult()

    local playerOdds
    local bankOdds

    for i = 1, self.player.numberOfUsedDecks, 1 do

        self.player:changeDeck(i)

        local result = Calculator:calculateResult(self.player:showDeck(), self.bank:showDeck())
        local playerMoney = self.player:showDeck().money

        if(result == ResultType.WIN) then
            print("You win!")
            playerOdds = 2
            bankOdds = -1
            self.winCount = self.winCount + 1
        elseif(result == ResultType.WINBYJACK) then
            print("BlackJack! You win!")
            playerOdds = 2.5
            bankOdds = -1.5
            self.winCount = self.winCount + 1
        elseif(result == ResultType.LOSEBYJACK) then
            print("You lost!")
            playerOdds = 0
            bankOdds = 1.5
            self.loseCount = self.loseCount + 1
        elseif(result == ResultType.TIE) then
            print("It's tie!")
            playerOdds = 1
            bankOdds = 1
        elseif(result == ResultType.LOSE) then
            print("You lost!")
            playerOdds = 0
            bankOdds = 1
            self.loseCount = self.loseCount + 1
        else print("error")
        end

        self.player:takeMoney(playerMoney * playerOdds)

        self.bank:takeMoney(playerMoney * bankOdds)

    end

    print("\nCurrent account: "..self.player:showMoney())

    local canPlay = true
    if(self.player:showMoney() <= 0) then canPlay = false end

    print("Continue? i - Yes \tq - No")
    local valid = false
    while(valid == false) do

        local key = inputHandler.readKey()
        if( key == ActionType.NEW)then
            if(canPlay) then
                self.state = GameState.RUNNING
                valid = true
            end
        elseif(key ==  ActionType.END) then
            self.state = GameState.END
            valid = true
        else
            valid = false
        end
    end


end

-- Interface functions

function Table:noticeUpdate()
    self.tableView:update()
end

function Table:noticeNewGame()
    self.state = GameState.NEW
end

function Table:noticeEndGame()
    self.state = GameState.END
end

function Table:noticeLowDeck()
    self.isLowDeck = true
end

function Table:noticePause()
    self.state = GameState.PAUSED
end

return Table

File: Kotlin/BlackJack/src/BlackJack/entity/Table.kt
package BlackJack.entity

import BlackJack.behavior.ClassicBank
import BlackJack.behavior.ClassicPlayer
import BlackJack.input.InputHandler
import BlackJack.observer.DealerObserver
import BlackJack.observer.PlayerObserver
import BlackJack.type.ActionType
import BlackJack.type.GameState
import BlackJack.type.ResultType
import BlackJack.gfx.ClassicTableView
import java.lang.Math.abs

class Table(numberOfDecks: Int, private val inputHandler: InputHandler, testing: Boolean = false) : DealerObserver, PlayerObserver {

    val dealer = if(!testing) Dealer(numberOfDecks, this) else TesterDealer(numberOfDecks, this)

    private val player = Player(0.0, dealer, ClassicPlayer(), inputHandler, this)

    private val bank = Bank(0.0, dealer, ClassicBank(), this)

    private var state = GameState.NEW

    private var isLowDeck = false

    private val tableView = ClassicTableView(player, bank)

    var winCount = 0
    private set
    var loseCount = 0
    private set


    fun startGame(){

        println("\nKöszöntjük a BlackJack asztalnál! Ne feledje, a bank mindig nyer :)\n")

        while (state != GameState.QUIT){

            when(state){

                GameState.RUNNING -> play()
                GameState.NEW -> settings()
                GameState.END -> showStatistics()
                GameState.QUIT -> {}
            }

        }
        println("Köszönjük, hogy a mi BlackJack alkalmazásunkat választotta! \nBye-bye! Have a good day!")
    }

    private  fun settings(){

        println("Adja meg az összeget, amivel játszani szeretne!")

        var newAmount = inputHandler.readNumber()

        while (newAmount == null) {
            println("Hibás bemenet!")
            newAmount = inputHandler.readNumber()
        }

        player.setBasicMoney(newAmount)

        winCount = 0
        loseCount = 0

        println("n - Start  q - Kilépés")
        var valid = false
        while(!valid){

            when(inputHandler.readKey()){

                ActionType.NEW -> {
                    state = GameState.RUNNING
                    valid = true
                }
                ActionType.END -> {
                    state = GameState.QUIT
                    valid = true
                }
                else -> valid = false
            }
        }
    }

    private fun showStatistics(){

        val coins = if(bank.showMoney() == 0.0) bank.showMoney() * -1 else bank.showMoney() * -1


        println("\nEredmények: \n" +
                "\tÖn enniyszer nyert: " + winCount.toString() + "\n" +
                "\tÖn enniyszer vesztett: " + loseCount.toString() + "\n" +
                "\tAz Ön nyeresége " + coins.toString() + " zseton.\n")

        println("n - Új játék  q - kilépés")

        var valid = false
        while(!valid){

            when(inputHandler.readKey()){

                ActionType.NEW -> {
                    state = GameState.NEW
                    valid = true
                }
                ActionType.END -> {
                    state = GameState.QUIT
                    valid = true
                }
                else -> valid = false
            }
        }

    }

    private fun play(){

        if(isLowDeck){

            dealer.createNewDeck()
            isLowDeck = false

        }

        player.preparation()
        bank.preparation()

        player.raise()

        player.receiveFirstCards()
        bank.receiveFirstCards()


        player.play()

        if( state != GameState.RUNNING ) return

        bank.play()

        calculateResult()

    }

    private fun calculateResult() {

        var playerOdds: Double
        var bankOdds: Double

        (0 until player.numberOfUsedDecks).forEach {

            player.changeDeck(it)

            val result = Calculator.calculateResult(player.showDeck(), bank.showDeck())
            val playerMoney = player.showDeck().money

            when(result){
                ResultType.WIN -> {
                    print("Ön nyert!")
                    playerOdds = 2.0
                    bankOdds = -1.0
                    winCount++
                }
                ResultType.WINBYJACK -> {
                    print("BlackJack! Ön nyert! ")
                    playerOdds = 2.5
                    bankOdds = -1.5
                    winCount++
                }
                ResultType.TIE -> {
                    print("Döntetlen! ")
                    playerOdds = 1.0
                    bankOdds = 1.0
                }
                ResultType.LOSE -> {
                    print("Ön vesztett! ")
                    playerOdds = 0.0
                    bankOdds = 1.0
                    loseCount++
                }
                ResultType.LOSEBYJACK -> {
                    print("Ön vesztett! ")
                    playerOdds = -0.5
                    bankOdds = 1.5
                    loseCount++
                }
            }

            if(player.showMoney() >= abs(playerMoney * playerOdds))
                player.takeMoney(playerMoney * playerOdds)
            else
                player.takeMoney(player.showMoney() * -1)


            bank.takeMoney(playerMoney * bankOdds)
        }

        println("\nJelenlegi összeg: " + player.showMoney())

        var canPlay = true
        if(player.showMoney() <= 0.0)
            canPlay = false

        println("Folytatja? i - Igen    q - Nem ")

        var valid = false
        while(!valid){

            when(inputHandler.readKey()){

                ActionType.NEW -> {

                    if(canPlay) {
                        state = GameState.RUNNING
                        valid = true
                    }
                }
                ActionType.END -> {
                    state = GameState.END
                    valid = true
                }
                else -> valid = false
            }
        }

    }

    fun getPlayerMoney() = player.showMoney()

    override fun noticeUpdate() { tableView.update() }

    override fun noticeNewGame() {state = GameState.NEW }

    override fun noticeEndGame() { state = GameState.END }

    override fun noticeLowDeck() { isLowDeck = true }
}